source/SpinalHDL/Libraries/Pipeline/introduction.rst

```diff
@@ -0,0 +1,794 @@
+    import spinal.lib.misc.pipeline._
+
+    class TopLevel extends Component {
+      val VALUE = SignalKey(UInt(16 bits))
+
+      val io = new Bundle{
+        val up = slave Stream(VALUE)  //VALUE can also be used as a HardType
+        val down = master Stream(VALUE)
+      }
+      
+      // NodesBuilder will be used to register all the nodes created, connect them via stages and generate the hardware
+      val builder = new NodesBuilder()
+
+      // Let's define a Node which connect from io.up
+      val n0 = new builder.Node{
+        arbitrateFrom(io.up)
+        VALUE := io.up.payload
+      }
+
+      // Let's define a Node which do some processing
+      val n1 = new builder.Node{
+        val RESULT = insert(VALUE + 0x1200)
+      }
+
+      //  Let's define a Node which connect to io.down
+      val n2 = new builder.Node {
+        arbitrateTo(io.down)
+        io.down.payload := n1.RESULT
+      }
+
+      // Let's connect those nodes by using registers stages and generate the related hardware
+      builder.genStagedPipeline()
+    }
+
+SignalKey
+============
+
+SignalKey class can be instanciated to represent some data which can go through the pipeline. Technicaly speaking, SignalKey is a HardType which has a name and is used as a "key" to retrieve stuff.
+
+.. code-block:: scala
+    
+    val PC = SignalKey(UInt(32 bits))
+    val PC_PLUS_4 = SignalKey(UInt(32 bits))
+
+    val n0, n1 = Node()
+    val s01 = StageConnector(n0, n1)
+
+    n0(PC) := 0x42
+    n1(PC_PLUS_4) := n1(PC) + 4
+
+Note that I got used to name the SignalKey instances using uppercase. This is to make it very explicit that the thing isn't a hardware signal, but are more like a "key/type" to access things.
+
+Node
+============
+
+Node mostly host the valid/ready arbitration signal, and the hardware signal required for all the SignalKey values going through it.
+
+You can access its arbitration via :
+
+
+.. list-table::
+   :header-rows: 1
+   :widths: 2 1 10
+
+   * - API
+     - Access
+     - Description
+   * - node.valid
+     - RW
+     - Is the signal which specify if a transaction is present on the node. It is driven by the upstream. Once asserted, it can only be dropped the cycle after which ready is high or node.isRemoved.
+   * - node.ready
+     - RW
+     - Is the signal which specify if the node's transaction should move away. It is driven by the downstream to create backpresure. The signal has no meaning when there is no transaction (node.valid being deasserted)
+   * - node.isValid
+     - RO
+     - node.valid's read only accessor
+   * - node.isReady
+     - RO
+     - node.ready's read only accessor
+   * - node.isFiring
+     - RO
+     - True when the node transaction is successfuly moving futher (isValid && isReady && !isRemoved). Useful to commit state changes.
+   * - node.isMoving
+     - RO
+     - True when the node transaction is moving away from the node (will not be in the node anymore starting from the next cycle),
```

Comment:
isMoving is also true when valid && !ready && isRemoved, so by moving away, i meant that it may moved downstream, but it may aswel move into void land.